File: src/lib/store.ts
import crypto from "node:crypto";
import { getDb } from "./db";
import { CODE_LENGTH, TTL_MS } from "./constants";
import type { ConfigRecord, ConfigStatus } from "./types";

/** Generate a cryptographically-random N-digit numeric code (e.g. "69168"). */
function generateCode(): string {
  const max = 10 ** CODE_LENGTH;
  const n = crypto.randomInt(0, max);
  return n.toString().padStart(CODE_LENGTH, "0");
}

/** Compute the live status of a record based on the current time. */
export function resolveStatus(record: ConfigRecord, now = Date.now()): ConfigStatus {
  return now >= record.expiresAt ? "expired" : record.status === "expired" ? "expired" : "active";
}

function buildShareLink(code: string): string {
  const base = (process.env.NEXT_PUBLIC_BASE_URL ?? "").replace(/\/$/, "");
  return `${base}/s/${code}`;
}

export interface CreatedConfig {
  record: ConfigRecord;
}

/**
 * Persist a new config, generating a guaranteed-unique 5-digit code.
 * Retries on the (rare) collision until it finds a free code.
 */
export function createConfig(config: string, baseUrlFromRequest?: string): ConfigRecord {
  const db = getDb();
  const insert = db.prepare(
    `INSERT INTO configs (id, code, config, shareLink, createdAt, expiresAt, status)
     VALUES (@id, @code, @config, @shareLink, @createdAt, @expiresAt, @status)`
  );

  const now = Date.now();
  const expiresAt = now + TTL_MS;

  // Try until a unique code is inserted. The UNIQUE index guarantees
  // correctness even under concurrency.
  for (let attempt = 0; attempt < 25; attempt += 1) {
    const code = generateCode();
    const base = (baseUrlFromRequest ?? process.env.NEXT_PUBLIC_BASE_URL ?? "").replace(/\/$/, "");
    const shareLink = base ? `${base}/s/${code}` : buildShareLink(code);

    const record: ConfigRecord = {
      id: crypto.randomUUID(),
      code,
      config,
      shareLink,
      createdAt: now,
      expiresAt,
      status: "active",
    };

    try {
      insert.run(record);
      return record;
    } catch (err) {
      // SQLITE_CONSTRAINT_UNIQUE -> duplicate code, retry.
      if (err instanceof Error && /UNIQUE/i.test(err.message)) {
        continue;
      }
      throw err;
    }
  }

  throw new Error("عدم امکان تولید کد یکتا. لطفاً دوباره تلاش کنید.");
}

/** Fetch a record by its code, refreshing its status if it has expired. */
export function getConfigByCode(code: string): ConfigRecord | null {
  const db = getDb();
  const row = db.prepare("SELECT * FROM configs WHERE code = ?").get(code) as
    | ConfigRecord
    | undefined;

  if (!row) return null;

  const status = resolveStatus(row);
  if (status !== row.status) {
    db.prepare("UPDATE configs SET status = ? WHERE code = ?").run(status, code);
    row.status = status;
  }

  return row;
}

/** Delete a record. Returns true when a row was actually removed. */
export function deleteConfigByCode(code: string): boolean {
  const db = getDb();
  const result = db.prepare("DELETE FROM configs WHERE code = ?").run(code);
  return result.changes > 0;
}

/** Housekeeping: mark all past-due records as expired. */
export function expireStaleConfigs(): void {
  const db = getDb();
  db.prepare("UPDATE configs SET status = 'expired' WHERE expiresAt <= ? AND status != 'expired'").run(
    Date.now()
  );
}

File: app/store.py
"""Data-access layer for config records."""
from __future__ import annotations

import secrets
import sqlite3
import time
import uuid
from dataclasses import dataclass
from typing import Optional

from . import config
from .db import get_db


@dataclass
class ConfigRecord:
    id: str
    code: str
    config: str
    shareLink: str
    createdAt: int
    expiresAt: int
    status: str

    @classmethod
    def from_row(cls, row: sqlite3.Row) -> "ConfigRecord":
        return cls(
            id=row["id"],
            code=row["code"],
            config=row["config"],
            shareLink=row["shareLink"],
            createdAt=row["createdAt"],
            expiresAt=row["expiresAt"],
            status=row["status"],
        )


def _now_ms() -> int:
    return int(time.time() * 1000)


def generate_code() -> str:
    """Cryptographically-random N-digit numeric code, e.g. '69168'."""
    upper = 10 ** config.CODE_LENGTH
    return str(secrets.randbelow(upper)).zfill(config.CODE_LENGTH)


def _resolve_status(record: ConfigRecord, now: Optional[int] = None) -> str:
    now = now if now is not None else _now_ms()
    if now >= record.expiresAt or record.status == "expired":
        return "expired"
    return "active"


def build_share_link(code: str, base_url: str = "") -> str:
    base = (base_url or config.BASE_URL).rstrip("/")
    return f"{base}/s/{code}" if base else f"/s/{code}"


def create_config(config_text: str, base_url: str = "") -> ConfigRecord:
    """Persist a new config with a guaranteed-unique 5-digit code."""
    conn = get_db()
    now = _now_ms()
    expires_at = now + config.TTL_SECONDS * 1000

    for _ in range(25):
        code = generate_code()
        record = ConfigRecord(
            id=str(uuid.uuid4()),
            code=code,
            config=config_text,
            shareLink=build_share_link(code, base_url),
            createdAt=now,
            expiresAt=expires_at,
            status="active",
        )
        try:
            conn.execute(
                """
                INSERT INTO configs (id, code, config, shareLink, createdAt, expiresAt, status)
                VALUES (:id, :code, :config, :shareLink, :createdAt, :expiresAt, :status)
                """,
                record.__dict__,
            )
            conn.commit()
            return record
        except sqlite3.IntegrityError:
            # Duplicate code (UNIQUE index) -> retry with a new one.
            continue

    raise RuntimeError("عدم امکان تولید کد یکتا. لطفاً دوباره تلاش کنید.")


def get_config_by_code(code: str) -> Optional[ConfigRecord]:
    """Fetch a record and refresh its status if it has expired."""
    conn = get_db()
    row = conn.execute("SELECT * FROM configs WHERE code = ?", (code,)).fetchone()
    if row is None:
        return None

    record = ConfigRecord.from_row(row)
    status = _resolve_status(record)
    if status != record.status:
        conn.execute("UPDATE configs SET status = ? WHERE code = ?", (status, code))
        conn.commit()
        record.status = status
    return record


def delete_config_by_code(code: str) -> bool:
    """Delete a record. Returns True when a row was removed."""
    conn = get_db()
    cur = conn.execute("DELETE FROM configs WHERE code = ?", (code,))
    conn.commit()
    return cur.rowcount > 0
